parquet-variant/src/main/java/org/apache/parquet/variant/Variant.java

```diff
@@ -0,0 +1,313 @@
+
+    // There is currently only one allowed version.
+    if ((metadata.get(metadata.position()) & VariantUtil.VERSION_MASK) != VariantUtil.VERSION) {
+      throw new UnsupportedOperationException(String.format(
+          "Unsupported variant metadata version: %d",
+          metadata.get(metadata.position()) & VariantUtil.VERSION_MASK));
+    }
+  }
+
+  /**
+   * @return the boolean value
+   */
+  public boolean getBoolean() {
+    return VariantUtil.getBoolean(value);
+  }
+
+  /**
+   * @return the byte value
+   */
+  public byte getByte() {
+    long longValue = VariantUtil.getLong(value);
+    if (longValue < Byte.MIN_VALUE || longValue > Byte.MAX_VALUE) {
+      throw new IllegalStateException("Value out of range for byte: " + longValue);
+    }
+    return (byte) longValue;
+  }
+
+  /**
+   * @return the short value
+   */
+  public short getShort() {
+    long longValue = VariantUtil.getLong(value);
+    if (longValue < Short.MIN_VALUE || longValue > Short.MAX_VALUE) {
+      throw new IllegalStateException("Value out of range for short: " + longValue);
+    }
+    return (short) longValue;
+  }
+
+  /**
+   * @return the int value
+   */
+  public int getInt() {
+    long longValue = VariantUtil.getLong(value);
+    if (longValue < Integer.MIN_VALUE || longValue > Integer.MAX_VALUE) {
+      throw new IllegalStateException("Value out of range for int: " + longValue);
+    }
+    return (int) longValue;
+  }
+
+  /**
+   * @return the long value
+   */
+  public long getLong() {
+    return VariantUtil.getLong(value);
+  }
+
+  /**
+   * @return the double value
+   */
+  public double getDouble() {
+    return VariantUtil.getDouble(value);
+  }
+
+  /**
+   * @return the decimal value
+   */
+  public BigDecimal getDecimal() {
+    return VariantUtil.getDecimal(value);
+  }
+
+  /**
+   * @return the float value
+   */
+  public float getFloat() {
+    return VariantUtil.getFloat(value);
+  }
+
+  /**
+   * @return the binary value
+   */
+  public ByteBuffer getBinary() {
+    return VariantUtil.getBinary(value);
+  }
+
+  /**
+   * @return the UUID value
+   */
+  public UUID getUUID() {
+    return VariantUtil.getUUID(value);
+  }
+
+  /**
+   * @return the string value
+   */
+  public String getString() {
+    return VariantUtil.getString(value);
+  }
+
+  /**
+   * The value type of Variant value. It is determined by the header byte.
+   */
+  public enum Type {
+    OBJECT,
+    ARRAY,
+    NULL,
+    BOOLEAN,
+    BYTE,
+    SHORT,
+    INT,
+    LONG,
+    STRING,
+    DOUBLE,
+    DECIMAL4,
+    DECIMAL8,
+    DECIMAL16,
+    DATE,
+    TIMESTAMP_TZ,
+    TIMESTAMP_NTZ,
+    FLOAT,
+    BINARY,
+    TIME,
+    TIMESTAMP_NANOS,
+    TIMESTAMP_NANOS_NTZ,
+    UUID
+  }
+
+  /**
+   * @return the type of the variant value
+   */
+  public Type getType() {
+    return VariantUtil.getType(value);
+  }
+
+  /**
+   * @return the number of object fields in the variant. `getType()` must be `Type.OBJECT`.
```

Comment:
Is the note here on getType because this is unchecked?

Reply:
What is the behavior if the variant is not an object?